library/pod-security-policy/users/src.rego

```diff
@@ -22,31 +42,76 @@ accept_users("MustRunAsNonRoot", provided_user) = res {res := provided_user != 0
 
 accept_users("MustRunAs", provided_user) = res  {
   ranges := input.parameters.runAsUser.ranges
-  matching := {1 | provided_user >= ranges[j].min; provided_user <= ranges[j].max}
-  res := count(matching) > 0
+  res := is_in_range(provided_user, ranges)
 }
 
-input_containers[c] {
-  c := input.review.object.spec.containers[_]
+# Group Options
+get_violation(field, params, container) = msg {
+  rule := params.rule
+  provided_value := get_field_value(field, container, input.review)
+  not is_null(provided_value)
+  not accept_value(rule, provided_value, params.ranges)
+  msg := sprintf("Container %v is attempting to run as disallowed group %v. Allowed %v: %v", [container.name, provided_value, field, params])
 }
 
-input_containers[c] {
-  c := input.review.object.spec.initContainers[_]
+get_violation(field, params, container) = msg {
+  is_null(get_field_value(field, container, input.review))
+  params.rule == "MustRunAs"
+  msg := sprintf("Container %v is attempting to run without a required securityContext/%v. Allowed %v: %v", [container.name, field, field, params])
 }
 
-run_as_user(container, review) = run_as_user {
-  has_field(container, "securityContext")
-  run_as_user := container.securityContext.runAsUser
+accept_value("RunAsAny", provided_value, ranges) {true}
+
+accept_value("MayRunAs", provided_value, ranges) = res { res := is_in_range(provided_value, ranges)}
+
+accept_value("MustRunAs", provided_value, ranges) = res { res := is_in_range(provided_value, ranges)}
+
+
+# If container level is provided, that takes precedence
+get_field_value(field, container, review) = out {
+  container_value := get_seccontext_field(field, container)
+  not is_null(container_value)
+  out := container_value
 }
 
-run_as_user(container, review) = run_as_user {
-  has_field(container, "securityContext")
-  not container.securityContext.runAsUser
+# If no container level exists, use pod level
+get_field_value(field, container, review) = out {
+  container_value := get_seccontext_field(field, container)
+  is_null(container_value)
   review.kind.kind == "Pod"
-  has_field(review.object.spec, "securityContext")
-  run_as_user := review.object.spec.securityContext.runAsUser
+  pod_value := get_seccontext_field(field, review.object.spec)
+  out := pod_value
+}
+
+# Helper Functions
+is_in_range(val, ranges) = res {
+  matching := {1 | val >= ranges[j].min; val <= ranges[j].max}
+  res := count(matching) > 0
 }
 
+# has_field returns whether an object has a field
 has_field(object, field) = true {
-    object[field]
+  object[field]
```

Comment:
Note that this will return `undefined` if an object has a field valued `false`, I don't think that matters in this case though.

Reply:
That's a good point. I don't think this function is necessary in this policy so I've pulled it out to be safe.